resin/logs.py

```diff
@@ -1,133 +1,194 @@
-from pubnub import Pubnub
 from functools import wraps
-import logging
+import json
+try:  # Python 3 imports
+    from urllib.parse import urljoin
+except ImportError:  # Python 2 imports
+    from urlparse import urljoin
+
+from twisted.internet import reactor, ssl
+from twisted.internet.defer import Deferred
+from twisted.internet.protocol import Protocol
+from twisted.web.client import Agent
+from twisted.web.http_headers import Headers
+from twisted.internet.ssl import ClientContextFactory
+
+from threading import Thread
 
 from .base_request import BaseRequest
+from . import exceptions
 from .models.config import Config
 from .models.device import Device
+from .settings import Settings
 
 
-logging.basicConfig(format='%(levelname)s:%(message)s')
+class WebClientContextFactory(ClientContextFactory):
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
+    def getContext(self, hostname, port):
+        return ClientContextFactory.getContext(self)
 
-def _print_deprecation_warning():
-    logging.warning(" using legacy logging services, this will stop working shortly.\nPlease update to ensure logs are correctly retrieved in future.")
 
+class StreamingParser(Protocol):
+    """
+    This is low level class and is not meant to be used by end users directly.
+    """
 
-# TODO: https://github.com/resin-io/resin-sdk/pull/277/files
-class Logs(object):
+    def __init__(self, callback, error):
+        self.callback = callback
+        self.error = error
+
+    def dataReceived(self, data):
+        d = Deferred()
+        d.addCallback(self.display_data)
+        d.addErrback(self.error_func)
+        d.callback(data)
+
+    def connectionLost(self, reason):
+        pass
+
+    def display_data(self, data):
+        lines = data.splitlines()
```

Comment:
this will drop logs if `data` contains a half finished line like this

```python
data = '{"message": "foobar"}\n{"mess'

Reply:
updated the behavior here, so user's error callback will be called in case of data loss or half finished data received (in this case, `{"message": "foobar"}` will be passed to user's callback and Failure contains ValueError exception will be passed to the error callback )

However, I don't have a strong opinion  here since I'm not so sure about the chunk sent by the API so if you think it's better to have a pending that stores unfinished log which will be merged with next chunk then I will update this.

Reply:
of course we should keep the pending data. Why would any user care that the line was received in two chunks?